onnxruntime/core/providers/xnnpack/nn/max_pool.cc

```diff
@@ -0,0 +1,193 @@
+
+  // use do {} while(false) so it's easier to set a breakpoint on the return
+  do {
+    // MaxPool has 1 input.
+    auto input_defs = node.InputDefs();
+    const auto& x_arg = *input_defs[0];
+
+    // we only support float currently
+    const auto* x_type = x_arg.TypeAsProto();
+    if (x_type == nullptr ||
+        x_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT) {
+      break;
+    }
+
+    // we only support 2D (4 dims with batch and channel)
+    const auto* x_shape = x_arg.Shape();
+    if (!x_shape || x_shape->dim_size() != 4) {
+      break;
+    }
+
+    // require C, H, W to be known so we can construct the xnnpack kernel prior to Compute
+    if (!x_shape->dim(1).has_dim_value() ||
+        !x_shape->dim(2).has_dim_value() ||
+        !x_shape->dim(3).has_dim_value()) {
+      break;
+    }
+
+    // we don't support creating the optional 'I' output
+    const auto& output_defs = node.OutputDefs();
+    if (output_defs.size() == 2 && output_defs[1]->Exists()) {
+      break;
+    }
+
+    ProtoHelperNodeContext nc(node);
+    OpNodeProtoHelper info(&nc);
+    PoolAttributes pool_attrs(info, "MaxPool", node.SinceVersion());
+
+    // xnnpack doesn't appear to support using 'ceil' to calculate the output shape
+    // https://github.com/google/XNNPACK/blob/3caa8b9de973839afa1e2a1462ff356e6927a66b/src/operators/max-pooling-nhwc.c#L256
+    // calls compute_output_dimension but there's no ability to specify rounding that value up.
+    if (pool_attrs.ceil_mode != 0) {
+      break;
+    }
+
+    if (!IsPaddingTypeSupported(pool_attrs.auto_pad)) {
+      break;
+    }
+
+    if ((pool_attrs.kernel_shape.size() != 2) ||
+        (pool_attrs.kernel_shape[0] == 1 && pool_attrs.kernel_shape[1] == 1)) {
+      // XNNPack doesn't support 1x1 maxpool.
+      break;
+    }
+
+    supported = true;
+  } while (false);
+
+  return supported;
+}
+
+MaxPool::MaxPool(const OpKernelInfo& info)
+    : OpKernel(info),
+      pool_attrs_{info, "MaxPool", info.node().SinceVersion()} {
+  uint32_t input_padding_top = gsl::narrow<uint32_t>(pool_attrs_.pads[0]);
+  uint32_t input_padding_left = gsl::narrow<uint32_t>(pool_attrs_.pads[1]);
+  uint32_t input_padding_bottom = gsl::narrow<uint32_t>(pool_attrs_.pads[2]);
+  uint32_t input_padding_right = gsl::narrow<uint32_t>(pool_attrs_.pads[3]);
+
+  uint32_t pooling_height = gsl::narrow<uint32_t>(pool_attrs_.kernel_shape[0]);
+  uint32_t pooling_width = gsl::narrow<uint32_t>(pool_attrs_.kernel_shape[1]);
+  uint32_t stride_height = gsl::narrow<uint32_t>(pool_attrs_.strides[0]);
+  uint32_t stride_width = gsl::narrow<uint32_t>(pool_attrs_.strides[1]);
+  uint32_t dilation_height = gsl::narrow<uint32_t>(pool_attrs_.dilations[0]);
+  uint32_t dilation_width = gsl::narrow<uint32_t>(pool_attrs_.dilations[0]);
+
+  // get values from any fusion with an activation
+  std::string activation;
+  if (info.GetAttr<std::string>("activation", &activation).IsOK()) {
```

Comment:
just an alternative to limit the scope of the GetAttr output variable
```suggestion
  if (std::string activation; info.GetAttr<std::string>("activation", &activation).IsOK()) {
```